oqsprov/oqs_encode_key2any.c

```diff
@@ -566,89 +640,262 @@ static int oqsx_pki_priv_to_der(const void *vxkey, unsigned char **pder)
+                                            aString[j]->length);
+                            ASN1_OCTET_STRING_free(aString[j]);
+                            OPENSSL_cleanse(aType[j]->value.sequence->data,
+                                            aType[j]->value.sequence->length);
+                            if (j < i)
+                                OPENSSL_clear_free(temp[j], templen[j]);
+                        }
+
+                        if (sk_ASN1_TYPE_num(sk) != -1)
+                            sk_ASN1_TYPE_pop_free(sk, &ASN1_TYPE_free);
+                        else
+                            ASN1_TYPE_free(aType[i]);
+
+                        OPENSSL_free(aType);
+                        OPENSSL_free(aString);
+                        OPENSSL_free(temp);
+                        OPENSSL_free(templen);
+                        PKCS8_PRIV_KEY_INFO_free(p8inf_internal);
+                        OPENSSL_free(name);
+                        return -1;
+                    }
+                } else
+                    buflen = oqsxkey->privkeylen_cmp[i];
+            } else {
+                nid = OBJ_sn2nid(name);
+                buflen = oqsxkey->privkeylen_cmp[i] + oqsxkey->pubkeylen_cmp[i];
+            }
 
-    OPENSSL_secure_clear_free(buf, buflen);
+            buf = OPENSSL_secure_malloc(buflen);
+            if (get_oqsname_fromtls(name)
+                != 0) { // include pubkey in privkey for PQC
+                memcpy(buf, oqsxkey->comp_privkey[i],
+                       oqsxkey->privkeylen_cmp[i]);
+                memcpy(buf + oqsxkey->privkeylen_cmp[i],
+                       oqsxkey->comp_pubkey[i], oqsxkey->pubkeylen_cmp[i]);
+            } else {
+                memcpy(buf, oqsxkey->comp_privkey[i], buflen);
+            }
+
+            if (nid == EVP_PKEY_EC) {
+                version = V_ASN1_OBJECT;
+                pval = OBJ_nid2obj(
+                    oqsxkey->oqsx_provider_ctx.oqsx_evp_ctx->evp_info->nid);
+            } else {
+                version = V_ASN1_UNDEF;
+                pval = NULL;
+            }
+            if (!PKCS8_pkey_set0(p8inf_internal, OBJ_nid2obj(nid), 0, version,
+                                 pval, buf, buflen)) {
+                for (int j = 0; j <= i; j++) {
+                    OPENSSL_cleanse(aString[j]->data, aString[j]->length);
+                    ASN1_OCTET_STRING_free(aString[j]);
+                    OPENSSL_cleanse(aType[j]->value.sequence->data,
+                                    aType[j]->value.sequence->length);
+                    OPENSSL_clear_free(temp[j], templen[j]);
+                }
+
+                sk_ASN1_TYPE_pop_free(sk, &ASN1_TYPE_free);
+                OPENSSL_free(name);
+                OPENSSL_free(aType);
+                OPENSSL_free(aString);
+                OPENSSL_free(temp);
+                OPENSSL_free(templen);
+                OPENSSL_cleanse(buf, buflen);
```

Comment:
I'm supposing you are speaking about the OPENSSL_cleanse.
I cannot free buf at that point because buf is used inside p8inf_internal, so if I free it when I free p8inf_internal I would get a double free error.
So I'm cleansing it to remove the pkey data from memory and on the next line I'm doing [PKCS8_PRIV_KEY_INFO_free](https://www.openssl.org/docs/manmaster/man3/PKCS8_PRIV_KEY_INFO_free.html) to free the p8inf_internal which include the buf.